packages/vats/src/bridge.js

```diff
@@ -15,76 +14,144 @@ import { makeWithQueue } from '@agoric/internal/src/queue.js';
+ * @param {import('@agoric/zone').Zone} zone
+ */
+export const zoneScopedManager = zone => {
+  const makeScopedManager = zone.exoClass(
+    'BridgeScopedManager',
+    undefined,
+    /**
+     * @param {string} bridgeId
+     * @param {import('@agoric/internal/src/callback').Callback} toBridge
+     * @param {import('@agoric/internal/src/callback').Callback} [fromBridge]
+     */
+    (bridgeId, toBridge, fromBridge) => ({
+      bridgeId,
+      toBridge,
+      fromBridge,
+    }),
+    {
+      toBridge(obj) {
+        return cb.esend(this.state.toBridge, obj);
+      },
+      fromBridge(obj) {
+        // If no handler was set, this will fail
+        const {
+          bridgeId,
+          fromBridge = cb.undef(`${bridgeId} bridge handler`),
+        } = this.state;
+        return cb.esend(fromBridge, obj);
+      },
+      setHandler(newHandler) {
+        // setHandler could probably be on a separate facet to separate it
+        // from the toBridge and fromBridge powers, but it's easier to
+        // implement a set once check for the handler and pass a single
+        // object around.
+        !this.state.fromBridge ||
+          Fail`Bridge handler already set for ${this.state.bridgeId}`;
+        newHandler || Fail`Bridge handler required`;
+        this.state.fromBridge = cb.method(newHandler, 'fromBridge');
+      },
+    },
+  );
+  return makeScopedManager;
+};
+
+/**
+ * @param {import('@agoric/zone').Zone} zone
  * @param {<T>(target: Device<T>) => T} D The device sender
- * @param {Device<BridgeDevice>} bridgeDevice The bridge to manage
- * @returns {import('./types.js').BridgeManager} admin facet for this handler
  */
-export function makeBridgeManager(E, D, bridgeDevice) {
+export const zoneBridgeManager = (zone, D) => {
+  const makeScopedManager = zoneScopedManager(zone.subZone('ScopedManager'));
+
   /**
-   * @type {MapStore<string, import('./types.js').ScopedBridgeManager>}
+   * Create a handler that demuxes/muxes the bridge device by its first argument.
+   *
+   * @param {Device<BridgeDevice>} bridgeDevice The bridge to manage
+   * @returns {{
+   *   manager: import('./types.js').BridgeManager,
+   *   handler: { inbound(srcID: string, obj: unknown): void },
+   * }}
    */
-  const scopedManagers = makeScalarMapStore('bridgeId');
+  const makeBridgeManagerKit = zone.exoClassKit(
+    'BridgeManagerKit',
+    undefined,
+    bridgeDevice => ({
+      /** @type {MapStore<string, ReturnType<ReturnType<zoneScopedManager>>>} */
+      scopedManagers: zone.detached().mapStore('scopedManagers'),
+      bridgeDevice,
+    }),
+    {
+      manager: {
+        register(bridgeId, handler) {
+          !this.state.scopedManagers.has(bridgeId) ||
+            Fail`Scoped bridge manager already registered for ${bridgeId}`;
+          const toBridge = cb.method(this.facets.manager, 'outbound', bridgeId);
+          const firstHandler = handler && cb.method(handler, 'fromBridge');
```

Comment:
hm. I wonder if there's some way to use `keyof` and such to preserve types.

Reply:
I have types now.

Reply:
The indirection is now avoided since it was unnecessary for the API that the bridge manager already exposed.  It does remain for chain storage since that uses a closure directly as a callback.